sympy/matrices/matrices.py

```diff
@@ -3888,22 +3895,78 @@ def LUdecomposition(self, iszerofunc=_iszero):
         LUdecompositionFF
         LUsolve
         """
-        combined, p = self.LUdecomposition_Simple(iszerofunc=_iszero)
-        L = self.zeros(self.rows)
-        U = self.zeros(self.rows)
-        for i in range(self.rows):
-            for j in range(self.rows):
-                if i > j:
-                    L[i, j] = combined[i, j]
-                else:
-                    if i == j:
-                        L[i, i] = 1
-                    U[i, j] = combined[i, j]
+
+        combined, p = self.LUdecomposition_Simple(iszerofunc=iszerofunc,
+                                                  rankcheck=rankcheck)
+
+        # L is lower triangular self.rows x self.rows
+        # U is upper triangular self.rows x self.cols
+        # L has unit diagonal. For each column in combined, the subcolumn
+        # below the diagonal of combined is shared by L.
+        # If L has more columns than combined, then the remaining subcolumns
+        # below the diagonal of L are zero.
+        # The upper triangular portion of L and combined are equal.
+
+        def entry_L(i, j):
+            if i < j:
+                return S.Zero
+            elif i == j:
+                return S.One
+            return combined[i, j]
+
+        L = self._new(combined.rows, combined.rows, entry_L)
+
+        def entry_U(i, j):
+            return S.Zero if i > j else combined[i, j]
+
+        U = self._new(combined.rows, combined.cols, entry_U)
+
         return L, U, p
 
-    def LUdecomposition_Simple(self, iszerofunc=_iszero):
-        """Returns A comprised of L, U (L's diag entries are 1) and
-        p which is the list of the row swaps (in order).
+    def LUdecomposition_Simple(self,
+                               iszerofunc=_iszero,
+                               rankcheck=False):
+        """Compute an lu decomposition of m x n matrix A, where P*A = L*U
+
+        * L is m x m lower triangular with unit diagonal
+        * U is m x n upper triangular
+        * P is an m x m permutation matrix
+
+        Returns an m x n matrix lu, and an m element list perm where each
+        element of perm is a pair of row exchange indices.
+
+        The factors L and U are stored in lu as follows:
+        The subdiagonal elements of L are stored in the subdiagonal elements
+        of lu, that is lu[i, j] = L[i, j] whenever i > j.
+        The elements on the diagonal of L are all 1, and are not explicitly
+        stored.
+        U is stored in the upper triangular portion of lu, that is
+        lu[i ,j] = U[i, j] whenever i <= j.
+        The output matrix can be visualized as:
+        Matrix([
+        [u, u, u, u],
+        [l, u, u, u],
+        [l, l, u, u],
+        [l, l, l, u]])
```

Comment:
This is great :-).  If you put it as an indented paragraph with blank lines before and after, it will render in the documentation as code instead of one line mushed together.